sound/soc/sof/ipc4.c

```diff
@@ -0,0 +1,242 @@
+			     void *reply_data)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	int ret;
+
+	/* wait for DSP IPC completion */
+	ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
+				 msecs_to_jiffies(sdev->ipc_timeout));
+	if (ret == 0) {
+		dev_err(sdev->dev, "ipc timed out for header:0x%x extension:0x%x",
+			msg->header, msg->extension);
+		return -ETIMEDOUT;
+	} else if (msg->reply_error) {
+		dev_err(sdev->dev, "ipc error for msg 0x%x : 0x%x\n",
+			msg->header, msg->extension);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sof_ipc4_tx_message_unlocked(struct snd_sof_ipc *ipc, u32 header,
+					u32 extension, void *msg_data, size_t msg_bytes,
+					void *reply_data, size_t reply_bytes)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	struct snd_sof_ipc_msg *msg;
+	int ret;
+
+	if (ipc->disable_ipc_tx)
+		return -ENODEV;
+
+	/*
+	 * The spin-lock is also still needed to protect message objects against
+	 * other atomic contexts.
+	 */
+	spin_lock_irq(&sdev->ipc_lock);
+
+	/* initialise the message */
+	msg = &ipc->msg;
+
+	msg->header = header;
+	msg->extension = extension;
+	msg->msg_size = msg_bytes;
+	msg->reply_size = reply_bytes;
+	msg->reply_error = 0;
+
+	/* attach any data */
+	if (msg_bytes > msg->max_payload_size || reply_bytes > msg->max_payload_size)
+		return -EINVAL;
+
+	memcpy(msg->msg_data, msg_data, msg_bytes);
+
+	sdev->msg = msg;
+
+	ret = snd_sof_dsp_send_msg(sdev, msg);
+	/* Next reply that we receive will be related to this message */
+	if (!ret)
+		msg->ipc_complete = false;
+
+	spin_unlock_irq(&sdev->ipc_lock);
+
+	if (ret) {
+		dev_err_ratelimited(sdev->dev, "ipc tx failed with error %d", ret);
+		return ret;
+	}
+
+	/* now wait for completion */
+
+	return ipc4_tx_wait_done(ipc, msg, reply_data);
+}
+
+int sof_ipc4_tx_message(struct snd_sof_ipc *ipc, u32 header, u32 extension,
+			void *msg_data, size_t msg_bytes, void *reply_data,
+			size_t reply_bytes)
+{
+	int ret;
+
+	/* Serialise IPC TX */
+	mutex_lock(&ipc->tx_mutex);
+
+	ret = sof_ipc4_tx_message_unlocked(ipc, header, extension, msg_data, msg_bytes,
+					   reply_data, reply_bytes);
+
+	mutex_unlock(&ipc->tx_mutex);
+
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc4_tx_message);
+
+void snd_sof_ipc4_msgs_rx(struct snd_sof_dev *sdev, u32 msg, u32 msg_ext)
+{
+	int err;
+
+	if (!SOF_IPC4_GLB_NOTIFY_MSG_TYPE(msg))
+		return;
+
+	switch (SOF_IPC4_GLB_NOTIFY_TYPE(msg)) {
+	case SOF_IPC4_GLB_NOTIFY_FW_READY:
+		/* check for FW boot completion */
+		if (sdev->fw_state == SOF_FW_BOOT_IN_PROGRESS) {
+			err = sof_ops(sdev)->fw_ready(sdev, msg);
+			if (err < 0)
+				sdev->fw_state = SOF_FW_BOOT_READY_FAILED;
+			else
+				sdev->fw_state = SOF_FW_BOOT_COMPLETE;
+
+			/* wake up firmware loader */
+			wake_up(&sdev->boot_wait);
+		}
+
+		break;
+	default:
+		break;
+	}
+}
+EXPORT_SYMBOL(snd_sof_ipc4_msgs_rx);
+
+int sof_ipc4_init_msg_memory(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_ipc_msg *msg;
+
+	msg = &sdev->ipc->msg;
+
+	//TODO: get max_payload_size from firmware
```

Comment:
dont you get a checkpatch error for using // instead of /* */?

Reply:
> dont you get a checkpatch error for using // instead of /* */?

No, but I will revise it. Thanks.